diff --git a/kernel/Makefile b/kernel/Makefile
index c21e6e5e..456fcc04 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -140,11 +140,11 @@ $(info -- KSU_SUSFS: adding function 'static inline const struct cred *get_cred_
 GET_CRED_RCU = static inline const struct cred *get_cred_rcu(const struct cred *cred)\n\
 {\n\t\
 	struct cred *nonconst_cred = (struct cred *) cred;\n\t\
 	if (!cred)\n\t\t\
 		return NULL;\n\t\
-	if (!atomic_inc_not_zero(&nonconst_cred->usage))\n\t\t\
+	if (!atomic_long_inc_not_zero(&nonconst_cred->usage))\n\t\t\
 		return NULL;\n\t\
 	validate_creds(cred);\n\t\
 	return cred;\n\
 }\n
 $(shell sed -i '/^static inline void put_cred/i $(GET_CRED_RCU)' $(srctree)/include/linux/cred.h;)
