diff --git a/drivers/Kconfig b/drivers/Kconfig
index 43377a3ac..da5128990 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -223,6 +223,7 @@ source "drivers/mux/Kconfig"
 
 source "drivers/sensors/Kconfig"
 
 source "drivers/esoc/Kconfig"
 
+source "drivers/kernelsu/Kconfig"
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 4a067525b..d9d5779bf 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -189,5 +189,7 @@ obj-$(CONFIG_MULTIPLEXER)	+= mux/
 obj-$(CONFIG_SENSORS_SSC)	+= sensors/
 obj-$(CONFIG_ESOC)              += esoc/
 # GNSS driver
 obj-$(CONFIG_GNSS_SIRF)		+= gnsssirf/
 obj-$(CONFIG_GNSS)		+= gnss/
+
+obj-$(CONFIG_KSU) += kernelsu/
